### DISCUSSION TITLE
2d not axial affine

### DISCUSSION THREAD

**Discussion by lilipj** [Project-MONAI/MONAI#4731] [General]
Hi,
I've 2d slices nifti files, generated from 3d with nibabel. These files are not axial, I've sagittal or coronal data.
The 3d affine is correct. I would like to extract the 2d affine with the to_affine_nd() function.
I wonder if this function works when data are not in axial orientation, because it takes the top left 2x2 elements of the 3d affine,
but I think it's only for axial orientation. Is that correct ?
Thanks in advance

**Comment by wyli** [comment]
yes, that utility function always extracts the upper left 2x2, you can normalise the spatial orientation with `Orientationd` before the extraction. 
` https://github.com/Project-MONAI/MONAI/blob/422cc6dcf5e7ba5243d5d585d632a7d54e0622e6/monai/transforms/spatial/dictionary.py#L393-L395

**Reply by anais2390** [reply]
Hi @wyli, I have the same problem: does this mean that we can't use the Transfom `Orientationd` on 2d inputs unless the input is axial ?

**Reply by wyli** [reply]
could you please share a code example? otherwise I'm not sure what 'the same problem' is.

**Reply by anais2390** [reply]
Hi @wyli, sorry for the late response. 

Indeed, I experience some trouble using transforms that need the affine information (such as `Orientation` and `Spacing`) on **2d** nifti data extracted from 3d **sagittal or coronal** nifti volumes. 

Here are some details: 

**Data**:

- one 2d nifti file extracted from 3d coronal nifti volume: [patient_1.nii.gz](https://github.com/Project-MONAI/MONAI/files/9445902/patient_1.nii.gz)

**My code**: 

`train_dict = [{"image": os.path.join(path_data, patient_name)}]`

`# load and preproc`
`train_transforms = Compose([LoadImaged(keys='image'), `
                                                `EnsureChannelFirstd(keys='image')`, 
`DataStatsd(keys="image"), `
`Spacingd(keys="image", pixdim=(5, 5)) `
`])`

`# create a training data loader`
`train_ds = CacheDataset(data=train_dict, transform=train_transforms)`
`train_loader = DataLoader(train_ds, batch_size=1, shuffle=True, num_workers=1, pin_memory=torch.cuda.is_available())`

`check = first(train_loader)`
`print(check['image'].shape)`

**Expected Output**:
>Data statistics:
>Type: <class 'numpy.ndarray'> float32
>Shape: (1, 512, 512)
>Value range: (0.0, 2134.666748046875)
>torch.Size([1, 1, 77, 77, 1]) 


**Output**:

>Data statistics:
>Type: <class 'numpy.ndarray'> float32
>Shape: (1, 512, 512)
>Value range: (0.0, 2134.666748046875)

Traceback (most recent call last):
  File "\.conda\envs\monai09\lib\site-packages\monai\transforms\transform.py", line 89, in apply_transform
    return _apply_transform(transform, data, unpack_items)
  File "\.conda\envs\monai09\lib\site-packages\monai\transforms\transform.py", line 53, in _apply_transform
    return transform(parameters)
  File "\.conda\envs\monai09\lib\site-packages\monai\transforms\spatial\dictionary.py", line 534, in __call__
    dtype=dtype,
  File "\.conda\envs\monai09\lib\site-packages\monai\transforms\spatial\array.py", line 448, in __call__
    new_affine = zoom_affine(affine_, out_d, diagonal=self.diagonal)
  File "\.conda\envs\monai09\lib\site-packages\monai\data\utils.py", line 761, in zoom_affine
    zs = np.linalg.cholesky(rzs.T @ rzs).T
  File "<__array_function__ internals>", line 6, in cholesky
  File "\.conda\envs\monai09\lib\site-packages\numpy\linalg\linalg.py", line 763, in cholesky
    r = gufunc(a, signature=signature, extobj=extobj)
  File "\.conda\envs\monai09\lib\site-packages\numpy\linalg\linalg.py", line 91, in _raise_linalgerror_nonposdef
    raise LinAlgError("Matrix is not positive definite")
numpy.linalg.LinAlgError: Matrix is not positive definite
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "\.conda\envs\monai09\lib\site-packages\IPython\core\interactiveshell.py", line 3457, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-7-5e2acaa2cd06>", line 25, in <module>
    train_ds = CacheDataset(data=train_dict, transform=train_transforms)
  File ".conda\envs\monai09\lib\site-packages\monai\data\dataset.py", line 723, in __init__
    self.set_data(data)
  File "\.conda\envs\monai09\lib\site-packages\monai\data\dataset.py", line 748, in set_data
    self._cache = _compute_cache()
  File "\.conda\envs\monai09\lib\site-packages\monai\data\dataset.py", line 737, in _compute_cache
    return self._fill_cache()
  File "\.conda\envs\monai09\lib\site-packages\monai\data\dataset.py", line 761, in _fill_cache
    desc="Loading dataset",
  File "\.conda\envs\monai09\lib\site-packages\tqdm\std.py", line 1195, in __iter__
    for obj in iterable:
  File "\.conda\envs\monai09\lib\multiprocessing\pool.py", line 748, in next
    raise value
  File "\.conda\envs\monai09\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "\.conda\envs\monai09\lib\site-packages\monai\data\dataset.py", line 777, in _load_cache_item
    item = apply_transform(_xform, item)
  File "\.conda\envs\monai09\lib\site-packages\monai\transforms\transform.py", line 113, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.spatial.dictionary.Spacingd object at 0x000001F4EB2894C8>


**Investigations**

- **Without `Spacingd`**: works fine
- **With `Spacingd` and 2d data extracted from axial volumes**: works fine
- The problem seems to be caused by the wrong computation of the 2d affine in the fonction `to_affine_nd()` that is suited only for axial data. 

**My questions are** :

- Is my analyse seems correct to you ? 
- If yes, is it possible to adapt the `to_affine_nd()` fonction so its works fine with other orientation than axial ? 

Thank you very much for you help !

**EDIT**

**The correct nifti file is this one -->

[patient_1_slice.nii.gz](https://github.com/Project-MONAI/MONAI/files/9468269/patient_1_slice.nii.gz) **

**Reply by wyli** [reply]
thanks @lilipj, the analysis makes sense, I'm adding the metadata updates for the `SqueezeDim` transform so that we can void the `to_affine_nd` function in this use case. https://github.com/Project-MONAI/MONAI/pull/5041

**Reply by anais2390** [reply]
Hi @lilipj , thanks you for your answer !

In my exemple, I don't use the `SqueezeDim` transform, maybe another way to do is to improve the `to_affine_nd()` to deal with coronal and sagittal data ? Or maybe I missed something ?

**Reply by wyli** [reply]
I think it's better to explicitly convert from spatial 3D to 2D. `to_affine_nd` won't address the issue, it's just a utility function to normalize the input shapes.

**Reply by anais2390** [reply]
Ok, so it means we can't load directly 2D slices ?

**Reply by wyli** [reply]
That means we should be careful with spatial shape (256,256) and (256,256,1). The former is the supported 2d use case. The latter is in 3d and it should be explicitly converted to 2d using SqueezeDim if the intention is to do 2d analysis.

**Reply by anais2390** [reply]
I agree for the (256, 256, 1) case, but actually, the former case (256, 256) is the case I described in the previous post and where we could not use the Spacing et Orientation transform, isn't it ?

**Reply by wyli** [reply]
in my understanding, the file you uploaded `"patient_1.nii.gz"` is a 3d volume with shape (512,512,1), or do I misunderstand:
```py
>>> print(nibabel.load("patient_1.nii.gz").header)
<class 'nibabel.nifti1.Nifti1Header'> object, endian='<'
sizeof_hdr      : 348
data_type       : b''
db_name         : b''
extents         : 0
session_error   : 0
regular         : b'r'
dim_info        : 0
dim             : [  3 512 512   1   1   1   1   1]
intent_p1       : 0.0

```

**Reply by anais2390** [reply]
Oh you are right, I am sorry I uploaded the wrong file. Actually, the code I posted correspond to what append when I load a 2d file (512, 512), I will provide you with the good file !
Actually, uploading a 3d file (512, 512, 1) is the solution I found to avoid the problems with `to_nd_affine()`

[patient_1_slice.nii.gz](https://github.com/Project-MONAI/MONAI/files/9468247/patient_1_slice.nii.gz)

**Reply by wyli** [reply]
so this file is not in a good shape, and can't be read by 3d slicer/itk-snap:
![Screenshot 2022-09-01 at 09 58 08](https://user-images.githubusercontent.com/831580/187875026-6c6850e6-b4a5-4f2f-a990-dd11764983fc.png)

if it's generated automatically by MONAI components, I'm curious about how it's generated and it's probably a bug from MONAI...

**Reply by wyli** [reply]
for future reference, given the input slice (512, 512, 1) in 3D (patient_1.nii.gz), the 2d analysis could be done by:
```py
import torch

from monai.data import *
from monai.transforms import *
from monai.utils import first


def main():
    train_dict = [{"image": "patient_1.nii.gz"}]

    train_transforms = Compose(
        [
            LoadImaged(keys="image", image_only=True, ensure_channel_first=True, as_closest_canonical=True),
            SqueezeDimd(keys="image", dim=2, update_meta=True),
            Spacingd(keys="image", pixdim=(5, 5)),
        ]
    )

    train_ds = CacheDataset(data=train_dict, transform=train_transforms)
    train_loader = DataLoader(train_ds, batch_size=1, shuffle=True, pin_memory=torch.cuda.is_available())

    check = first(train_loader)
    print(check["image"].shape, check["image"].affine)


if __name__ == "__main__":
    main()

```

the `as_closest_canonical=True` will make it more robust when `SqueezeDim` drops any of the spatial dimensions.

**Reply by anais2390** [reply]
Thanks you @wyli !
The 2d file was not generated via MONAI but using nibabel librairy.
